File: TypeScript/src/tax_table.ts
import { BigDecimal } from 'bigdecimal';
// https://github.com/iriscouch/bigdecimal.js

class IncomeRange {
  public readonly threshold: BigDecimal;
  public readonly percentage: BigDecimal;

  constructor(threshold: BigDecimal, percentage: BigDecimal) {
    this.threshold = threshold;
    this.percentage = percentage;
  }
}

export class TaxTable {
  private noTax: BigDecimal;
  private levels: IncomeRange[];

  constructor() {
    this.noTax = new BigDecimal(0);
    this.levels = [];
  }

  addProgression(threshold: BigDecimal, percentage: number): void {
    const decimalPercentage = new BigDecimal(percentage).divide(new BigDecimal(100));
    this.levels.push(new IncomeRange(threshold, decimalPercentage));
  }

  taxFor(income: BigDecimal): BigDecimal {
    for (let i = 0; i < this.levels.length; i++) {
      if (income.compareTo(this.levels[i].threshold) >= 0) {
        return income.multiply(this.levels[i].percentage);
      }
    }
    return this.noTax;
  }

}

File: Java/src/main/java/gov/TaxTable.java
package gov;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

public class TaxTable {

    private final BigDecimal noTax = BigDecimal.ZERO;

    static class IncomeRange {
        public final BigDecimal threshold;
        public final BigDecimal percentage;

        public IncomeRange(BigDecimal threshold, BigDecimal percentage) {
            this.threshold = threshold;
            this.percentage = percentage;
        }
    }

    private final List<IncomeRange> levels = new ArrayList<>();

    public void addProgression(BigDecimal threshold, int percentage) {
        BigDecimal decimalPercentage = new BigDecimal(percentage)
                .divide(new BigDecimal(100));
        levels.add(new IncomeRange(threshold, decimalPercentage));
    }

    public BigDecimal taxFor(BigDecimal income) {
        for (IncomeRange level : levels) {
            if (income.compareTo(level.threshold) >= 0) {
                return income.multiply(level.percentage);
            }
        }
        return noTax;
    }

}
